feat: user auth implemented

amaazahmad opened the pull request.

amaazahmad committed:
TYPE: feature

SUBJECT: Basic components for Form + Routing for Login/signup

DESCRIPTION: InputBar and Form components have been created to be used in both login and signup forms. Default routing set to the login page using <Navigate/>. Routing between Signup and Login pages settled.

amaazahmad committed:
TYPE: Refactor

SUBJECT: Deleted unused files. Changed the method of accessing form elements in userForm.

DESCRIPTION: Deleted a few default files which were not being used e.g. webVitals, etc. Used the FormEvent instead of SyntheticEvent in userForm component and changed the method of accessing form properties.

amaazahmad committed:
TYPE: feature

SUBJECT: Firebase user authentication applied + HTML form replaced with react-hook-form

DESCRIPTION: Firebase authentication with email/password applied. Still not able to store the name of the user, though. The retrieved user object is yet to be stored in Context and LocalStorage. The HTML form element in userForm now uses react-hook-form.

amaazahmad committed:
TYPE: fix

SUBJECT: Password and Confirm Password validation added in react hook form.

amaazahmad committed:
TYPE: feature-improvement

SUBJECT: Signout functionality implemented. Routing based on user logged-in status.

DESCRIPTION: Signout functionality completely implemented. Didn't use local storage as Firebase keeps the track of logged-in user with auth object. Routing implemented: default page set to login if the user is not logged-in, default page is home if the user is logged-in.

amaazahmad committed:
TYPE: styles

SUBJECT: Login and Signup pages styling for mobile display

DESCRIPTION: Added styling for the login and signup pages by creating CSS files for now, with a fixed screen size in px. Responsiveness and Tailwind yet to be done.

amaazahmad committed:
TYPE: styling fix

SUBJECT: Placeholder vs text color fixed for input fields in login/signup forms.

amaazahmad committed:
TYPE: styles

SUBJECT: Responsive settled and login/signup screens made for larger screens

DESCRIPTION: Added a side-bar for login/signup form for larger screens. Container divs added in login and signup pages for CSS implementation.

YET TO BE DONE:
- Use Tailwind instead of CSS files
- Adjust the heading size in the left sidebar of login/signup page for medium sized screens. It stacks on different lines for med screens.

amaazahmad committed:
TYPE: Styling-fixes

SUBJECT: Styling fixes done

DESCRIPTION: Following styling fixes done in login/signup pages:
- Left aligned welcome heading and subheading for desktop screen
- Input fields max-width adjusted
- Form's padding adjusted
- Font-size readjusted for headings and input fields
- LeftImageContainer size increased from 33% to 37%

amaazahmad committed:
TYPE: Styling fixes

SUBJECT: Tiny styling improvements for responsiveness

amaazahmad committed:
TYPE: styling-fix

SUBJECT: Styling improved after first demo

DESCRIPTION:
- Showed user not found error instead of incorrect password error. FIXED
- Increased padding with left side of form: Changed to 8rem from 4rem after 1024px screen width
- Error font size slightly increased.

maryamrasool commented on src/App.tsx:
```diff
@@ -1,26 +1,35 @@
+		const path = currentUser ? "/" : "/login"
+		console.log("PATH CHANGED TO: ", path)
+		navigate(path)
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
Do not push commented code to remote repo.

amaazahmad replied:
<img width="469" alt="image" src="https://user-images.githubusercontent.com/111420831/191839949-586ce787-e8e4-4b4a-9e49-806de51b09d5.png">

It is meant for the linter to stop showing a warning for not including 'navigate' in the dependencies array. I tried to remove it, but it starts showing the warning.

maryamrasool commented on src/App.tsx:
```diff
@@ -1,26 +1,35 @@
+
+	useEffect(() => {
+		const path = currentUser ? "/" : "/login"
+		console.log("PATH CHANGED TO: ", path)
```
Do not push console log statements to remote repo.

amaazahmad replied:
My bad.
Removed.

maryamrasool commented on src/App.tsx:
```diff
@@ -1,26 +1,35 @@
+		const path = currentUser ? "/" : "/login"
+		console.log("PATH CHANGED TO: ", path)
+		navigate(path)
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
Do not push commented code to remote repo.

maryamrasool commented on src/App.tsx:
```diff
@@ -1,26 +1,35 @@
-import React from 'react';
-import logo from './logo.svg';
-import './App.css';
+import React, {useEffect} from "react"
+import "./App.css"
+import LoginPage from "./pages/login/login.component"
+import SignupPage from "./pages/signup/signup.component"
+import HomePage from "./pages/home/home.component"
+import {useNavigate} from "react-router-dom"
+
+import {useContext} from "react"
+import {UserContext} from "./context/user.context"
+
+import {Routes, Route} from "react-router-dom"
```
The import statements should be in an order. I suggest using the following order for imports:
- React Imports
- Third Party Packages
- Components
- Context
- Custom Hooks
- Utils
- Constants

Also, all imports from one package should be in one single line. For example on line 1 and line 8, you are importing items from the react package. This should be one import.

`import React, {useEffect, useContext} from "react"`

amaazahmad replied:
<img width="481" alt="image" src="https://user-images.githubusercontent.com/111420831/191844266-b078104a-0293-4efb-89b7-77288bdd3315.png">

Done. 
Although, kindly suggest if it is ok to add comments like shown in the above screenshot.

maryamrasool commented on src/components/inputBar/inputBar.component.tsx:
```diff
@@ -0,0 +1,17 @@
+	placeHolder: string;
+};
+
+function InputBar(props: InputBarProps) {
```
Instead of using these regular functions, it's better to use arrow functions. Arrow functions were introduced in ES6. You can read about them [here](https://betterprogramming.pub/difference-between-regular-functions-and-arrow-functions-f65639aba256).

amaazahmad replied:
Understood.

maryamrasool commented on src/components/inputBar/inputBar.component.tsx:
```diff
@@ -0,0 +1,17 @@
+	placeHolder: string;
+};
+
+function InputBar(props: InputBarProps) {
```
Props are not de structured inside the function. They are de structured in the function parameters. 
```suggestion
const InputBar = ({name, type, placeHolder}: InputBarProps) => {
```

maryamrasool commented on src/components/userForm/userForm.component.tsx:
```diff
@@ -0,0 +1,114 @@
+import {
+	createUserAccount,
+	signInUser,
+} from "../../utils/firebase/firebase.utils"
+import {useForm} from "react-hook-form"
+import {useNavigate} from "react-router-dom"
```
Same comment for organising your imports in the correct order.

amaazahmad replied:
Done.

maryamrasool commented on src/components/userForm/userForm.component.tsx:
```diff
@@ -0,0 +1,114 @@
+	isLogin: boolean
+}
+
+const UserForm = (props: UserFormProps) => {
```
De structure your props here.

amaazahmad replied:
Done.

maryamrasool commented on src/components/userForm/userForm.component.tsx:
```diff
@@ -0,0 +1,114 @@
+		register,
+		handleSubmit,
+		getValues,
+		//formState: {errors},
```
Do not push commented code to remote repo.

amaazahmad replied:
My bad. Missed it.

Removed.

maryamrasool commented on src/components/userForm/userForm.component.tsx:
```diff
@@ -0,0 +1,114 @@
+		const password = data.password_input
+		const confirmPassword = data.confirm_password_input
+
+		console.log(name, email, password, confirmPassword)
```
Do not push console log statements to remote repo.

amaazahmad replied:
Done.

maryamrasool commented on src/components/userForm/userForm.component.tsx:
```diff
@@ -0,0 +1,114 @@
+		const confirmPassword = data.confirm_password_input
+
+		console.log(name, email, password, confirmPassword)
+		let resp
+		if (isLogin) resp = await signInUser(email, password)
+		else resp = await createUserAccount(email, password)
+
+		if (!(resp instanceof Error)) navigate("/home")
```
Although this code works fine, it is not a good approach. Also, `resp` is not a good name for this variable. Instead, you could name this variable as `UserResponse`.

Suggestion:
```
const UserResponse = isLogin
      ? await signInUser(email, password)
      : await createUserAccount(email, password);

    if (!(UserResponse instanceof Error)) navigate("/home");
```

amaazahmad replied:
Fixed

maryamrasool commented on src/components/userForm/userForm.component.tsx:
```diff
@@ -0,0 +1,114 @@
+	})
+
+	return (
+		<div className="formContainer">
```
I know that you're going to remove this class name, but for your reference, class names should be in kebab case.

amaazahmad replied:
Understood.

maryamrasool commented on src/components/userForm/userForm.component.tsx:
```diff
@@ -0,0 +1,114 @@
+			<form className="formElement" onSubmit={onSubmitHandler}>
+				{!isLogin && (
+					<input
+						{...register("name_input", {required: "Name is required."})}
```
It's better to use camelCase for variables.

```suggestion
						{...register("nameInput", {required: "Name is required."})}
```
This goes for all the variables.

amaazahmad replied:
Fixed.

maryamrasool commented on src/components/userForm/userForm.component.tsx:
```diff
@@ -0,0 +1,114 @@
+	return (
+		<div className="formContainer">
+			<form className="formElement" onSubmit={onSubmitHandler}>
+				{!isLogin && (
```
Instead of using the && operator, it's better to use the ternary operator.

amaazahmad replied:
<img width="742" alt="Screenshot 2022-09-23 at 1 56 33 AM" src="https://user-images.githubusercontent.com/111420831/191849550-dab296e2-104a-4150-b670-3e857f38f6fa.png">

If I use the ternary operator, I have the pass in an empty Fragment as the first argument, is that ok?

amaazahmad replied:
I went ahead with your suggestion, though.

maryamrasool commented on src/components/userForm/userForm.component.tsx:
```diff
@@ -0,0 +1,114 @@
+						required: "Password is required.",
+						minLength: {
+							value: 6,
+							message: "Password must be atleast 6 characters long!",
```
Remove exclamation mark at the end of the error message. It makes the error message look aggressive.

amaazahmad replied:
Done

maryamrasool commented on src/components/userForm/userForm.component.tsx:
```diff
@@ -0,0 +1,114 @@
+				)}
+				<button type="submit">{isLogin ? "LOGIN" : "SUBMIT"}</button>
+			</form>
+			{/* <form onSubmit={onSubmitHandler}>
```
Do not push commented code to remote repo.

amaazahmad replied:
Fixed

maryamrasool commented on src/context/user.context.tsx:
```diff
@@ -0,0 +1,9 @@
+import firebase from "firebase/auth"
+import {createContext} from "react"
```
Make sure that react import is always on top.

amaazahmad replied:
Done.

maryamrasool commented on src/context/user.provider.tsx:
```diff
@@ -0,0 +1,26 @@
+import React from "react"
+import {useState, useEffect} from "react"
```
Imports from one package should be in one line.

```suggestion
import React, {useState, useEffect} from "react"
```

amaazahmad replied:
Done.

maryamrasool commented on src/context/user.provider.tsx:
```diff
@@ -0,0 +1,26 @@
+}
+
+export const UserProvider: React.FC<UserProviderProps> = (
+	props: UserProviderProps
```
De structure props here instead of doing it in the function

amaazahmad replied:
Done.

maryamrasool commented on src/context/user.provider.tsx:
```diff
@@ -0,0 +1,26 @@
+		return unsubscribeFunc
+	}, [])
+
+	return <UserContext.Provider value={value}>{children}</UserContext.Provider>
```
Instead of creating a constant for value up on line 16, it's better to send multiple values in the values attribute.
```suggestion
	return <UserContext.Provider value={{ currentUser, setCurrentUser }}>{children}</UserContext.Provider>
```
Remove the value constant at line 16, afterwards.

amaazahmad replied:
Done

maryamrasool commented on src/pages/home/home.component.tsx:
```diff
@@ -0,0 +1,19 @@
+import {signOutUser} from "../../utils/firebase/firebase.utils"
+import {useNavigate} from "react-router-dom"
```
Same comment for organising your imports.

amaazahmad replied:
Done.

maryamrasool commented on src/index.tsx:
```diff
@@ -1,19 +1,24 @@
-    <App />
-  </React.StrictMode>
-);
+	// <React.StrictMode>
```
Remove commented code.

amaazahmad replied:
Uncommented this.

I commented this for debugging.

maryamrasool commented on src/index.tsx:
```diff
@@ -1,19 +1,24 @@
+)
 
 // If you want to start measuring performance in your app, pass a function
 // to log results (for example: reportWebVitals(console.log))
 // or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
-reportWebVitals();
+// reportWebVitals();
```
Remove this commented code as well.

amaazahmad replied:
Done.

maryamrasool commented on src/pages/home/home.component.tsx:
```diff
@@ -0,0 +1,19 @@
+function HomePage() {
+	const navigate = useNavigate()
+	const signOutHandler = () => {
+		const resp = signOutUser()
```
It is not advised to use these kind of shortcuts for variable names. Better to rename it to `response`

amaazahmad replied:
Fixed

maryamrasool commented on src/pages/login/login.component.tsx:
```diff
@@ -0,0 +1,34 @@
+import "./login.styles.css"
+
+function LoginPage() {
+	console.log("LOGIN PAGE RENDERED")
```
Do not push console statements to remote repo.

amaazahmad replied:
Removed.

maryamrasool commented on src/pages/signup/signup.styles.css:
```diff
@@ -0,0 +1,39 @@
+/* .welcomeTextContainer {
```
Do not push commented code to remote.

amaazahmad replied:
CSS files removed altogether.

maryamrasool commented on src/utils/firebase/firebase.utils.ts:
```diff
@@ -0,0 +1,55 @@
+	signOut,
+} from "firebase/auth"
+const firebaseConfig = {
+	apiKey: "AIzaSyBhKb0MOulldYv0eQ8Znfr5-7ng_Sh5pXQ",
```
Do not push these important keys to remote repo ever. Create .env file and store these variables there. Import these values from there and use them.

amaazahmad replied:
Done.
Created .env.local file.

amaazahmad committed:
style: Tailwind CSS implemented

Description: All CSS files have been removed and tailwind has been implemented. An unused InputBar component has been deleted.

amaazahmad committed:
refactor: fixed issues raised in the first PR

Description:
- simple functions changed to arrow functions
- comments and log statements removed
- props destructred in the function definition
- imports sorted
- firebase config shifted to .env.local

amaazahmad committed:
fix: Fixed the flashing of login screen if a user is logged and refreshes the home page

description: added a onAuthStateChanged listener in the useEffect hook of the App.tsx. This reduces the delay caused by setting context n then reading the currentUser from there.

amaazahmad committed:
fix+refactor: Made the form component generic and Login screen flashing fixed

Description:
- The login screen used to flash for a second if a logged-in user refreshes the application. That was fixed by using the onAuthStateChanged listener directly in the useEffect hook.
- The form component was made generic and then used in both Login and Sign up pages.

amaazahmad committed:
refactor: minor improvements in error messages and data type for react-hook-form

description:
- Used the FieldValues data type in place of any type on the submitHandler function in react-hook-form
- Removed the validation rule for confirm password and password from the confirmPassword field. This is now done in onSubmitHandler
- Moved the errorMessage for firebase errors outside the form, in the login and signup components independently.

maryamrasool commented on src/App.tsx:
```diff
@@ -1,26 +1,44 @@
-import React from 'react';
-import logo from './logo.svg';
-import './App.css';
+// React Imports
```
You do not need to add comments like these. The imports are self explanatory and the organisation is done for code readability only.

amaazahmad replied:
Understood.

I've removed the comments.

maryamrasool commented on src/App.css:
```diff
@@ -1,38 +1,39 @@
 .App-logo {
-  height: 40vmin;
-  pointer-events: none;
+	height: 40vmin;
```
Why are we using vmin here?

amaazahmad replied:
Really have no idea...
This is the default App.css file.

maryamrasool commented on src/App.tsx:
```diff
@@ -1,26 +1,44 @@
+
+	useEffect(
+		() => {
+			// the onAuthStateChanged function is used here instead of getting the current user from context because it had a delay. The login screen flashed each time a logged-in user refreshed the home page. The usage of onAuthStateChanged fixes it.
```
I appreciate adding this comment here to let me know about this function. However, comments like these are not generally pushed to the remote repo.

amaazahmad replied:
This issue was solved after a lot of time n effort so I thought it'd be better to write a comment here...

I've removed it.

maryamrasool commented on src/components/form/form.component.tsx:
```diff
@@ -0,0 +1,88 @@
+					await onSubmitHandler(data)
+				})}
+			>
+				{fields.map((field: FormFields, index: number) => {
```
You do not need to add types for field and index here. The map function already knows that it will return an item from the fields array of type FormFields. 
Also, index is not used anywhere so you can remove it from here.
```suggestion
				{fields.map((field) => {
```

amaazahmad replied:
Understood n fixed.

maryamrasool commented on src/components/form/form.component.tsx:
```diff
@@ -0,0 +1,88 @@
+					}
+
+					return (
+						<React.Fragment key={field.name}>
```
I will suggest to import Fragment from react on the top and just use Fragment tag here.

`import React, { Fragment } from "react";`

```suggestion
						<Fragment key={field.name}>
```

amaazahmad replied:
Done

maryamrasool commented on src/components/form/form.component.tsx:
```diff
@@ -0,0 +1,88 @@
+					await onSubmitHandler(data)
+				})}
+			>
+				{fields.map((field: FormFields, index: number) => {
```
Instead of getting the whole field here, you can destruct the required items from it.
```suggestion
				{fields.map(({ name, type, minLength, pattern, required }) => {
```

amaazahmad replied:
My bad...
Done.

maryamrasool commented on src/components/form/form.component.tsx:
```diff
@@ -0,0 +1,88 @@
+                                   sm:self-start sm:ml-0 sm:text-sm sm:h-12
+                                   lg:h-[60px] lg:text-lg lg:max-w-[600px] "
+								type={field.type}
+								placeholder={
```
Do not add hard coded placeholders here. You should be passing the placeholder in the fields object from the parent component.

amaazahmad replied:
<img width="448" alt="Screenshot 2022-10-06 at 6 10 56 PM" src="https://user-images.githubusercontent.com/111420831/194321595-b8fa567d-6d38-4908-ab1b-1b1a06ff7b34.png">

The hard-coded place-holder is just for the confirmPassword field, actually. 
For all the other fields, using the name of the field with a "Enter your ... " string seemed fine.

But I do understand what you're suggesting.

maryamrasool commented on src/components/form/form.component.tsx:
```diff
@@ -0,0 +1,88 @@
+									pattern: {...patternObj},
+									minLength: {...minLengthObj},
+								})}
+							></input>
```
Where there is nothing to add in the body of the tag, make it self closing.
`<input ... />`

amaazahmad replied:
Understood. Done.

maryamrasool commented on src/context/user.provider.tsx:
```diff
@@ -0,0 +1,33 @@
+	children: React.ReactNode
+}
+
+export const UserProvider: React.FC<UserProviderProps> = ({
+	children,
+}: UserProviderProps) => {
```
If you've imported react on the top, you do not need to define the function like this.

```suggestion
export const UserProvider = ({ children }: UserProviderProps) => {
```

amaazahmad replied:
Done.

maryamrasool commented on src/index.tsx:
```diff
@@ -1,19 +1,23 @@
+	</React.StrictMode>
+)
 
 // If you want to start measuring performance in your app, pass a function
 // to log results (for example: reportWebVitals(console.log))
 // or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
```
These comments were related to analytics and reportWebVitals which you have already removed. Hence, you should also remove the comments related to it.

amaazahmad replied:
Done.

maryamrasool commented on src/pages/login/login.component.tsx:
```diff
@@ -0,0 +1,99 @@
+					</p>
+				</div>
+				<div className="w-full">
+					<Form
+						fields={LoginFormFields}
+						buttonText="LOGIN"
+						onSubmitHandler={onSubmitHandler}
+					></Form>
```
Since we're not going to pass anything in the Form component. Let's make it self closing tag.
```suggestion
					<Form
						fields={LoginFormFields}
						buttonText="LOGIN"
						onSubmitHandler={onSubmitHandler}
					/>
```

amaazahmad replied:
Done.

maryamrasool commented on src/pages/signup/signup.component.tsx:
```diff
@@ -0,0 +1,115 @@
+					</p>
+				</div>
+				<div className="w-full">
+					<Form
+						fields={SignupFormFields}
+						buttonText="SUBMIT"
+						onSubmitHandler={onSubmitHandler}
+					></Form>
```
Same comment for making the tag self closing.

amaazahmad replied:
Done.
